File: src/providers/blink.ts
import type { Invoice, LightningProvider } from "../types";

/**
 * Blink (blink.sv) — free custodial Lightning provider.
 * Get API key: dashboard.blink.sv → API Keys
 */
export class BlinkProvider implements LightningProvider {
  private apiKey: string;
  private walletId: string;

  constructor(apiKey: string, walletId: string) {
    this.apiKey = apiKey;
    this.walletId = walletId;
  }

  async createInvoice(amountSats: number): Promise<Invoice> {
    const res = await fetch("https://api.blink.sv/graphql", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-API-KEY": this.apiKey,
      },
      body: JSON.stringify({
        query: `mutation CreateInvoice($input: LnInvoiceCreateInput!) {
          lnInvoiceCreate(input: $input) {
            invoice { paymentRequest paymentHash }
            errors { message }
          }
        }`,
        variables: {
          input: {
            walletId: this.walletId,
            amount: amountSats,
            memo: "L402 API access",
          },
        },
      }),
    });

    if (!res.ok) {
      const err = await res.text().catch(() => res.statusText);
      throw new Error(`Blink invoice creation failed (${res.status}): ${err}`);
    }

    const json = (await res.json()) as {
      data: {
        lnInvoiceCreate: {
          invoice: { paymentRequest: string; paymentHash: string };
          errors: { message: string }[];
        };
      };
    };

    const { invoice, errors } = json.data.lnInvoiceCreate;
    if (errors?.length) throw new Error(`Blink error: ${errors[0].message}`);

    const macaroon = Buffer.from(JSON.stringify({
      hash: invoice.paymentHash,
      exp: Math.floor(Date.now() / 1000) + 3600,
    })).toString("base64");

    return {
      paymentRequest: invoice.paymentRequest,
      paymentHash: invoice.paymentHash,
      macaroon,
      amountSats,
      expiresAt: Math.floor(Date.now() / 1000) + 3600,
    };
  }

  async checkPayment(paymentHash: string): Promise<boolean> {
    const res = await fetch("https://api.blink.sv/graphql", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-API-KEY": this.apiKey,
      },
      body: JSON.stringify({
        query: `query CheckInvoice($paymentHash: PaymentHash!) {
          lnInvoice(paymentHash: $paymentHash) { status }
        }`,
        variables: { paymentHash },
      }),
    });
    if (!res.ok) return false;
    const json = (await res.json()) as { data: { lnInvoice: { status: string } } };
    return json.data?.lnInvoice?.status === "PAID";
  }
}

File: python/l402kit/providers/blink.py
import time
import base64
import json
import httpx
from ..types import Invoice


class BlinkProvider:
    """
    Blink Lightning provider — free, no node needed.
    Get API key at dashboard.blink.sv
    """

    def __init__(self, api_key: str, wallet_id: str):
        self.api_key = api_key
        self.wallet_id = wallet_id

    async def create_invoice(self, amount_sats: int) -> Invoice:
        query = """
        mutation CreateInvoice($input: LnInvoiceCreateInput!) {
          lnInvoiceCreate(input: $input) {
            invoice { paymentRequest paymentHash }
            errors { message }
          }
        }
        """
        async with httpx.AsyncClient() as client:
            res = await client.post(
                "https://api.blink.sv/graphql",
                headers={"X-API-KEY": self.api_key},
                json={"query": query, "variables": {
                    "input": {"walletId": self.wallet_id, "amount": amount_sats, "memo": "L402 API access"}
                }},
            )
            res.raise_for_status()
            data = res.json()["data"]["lnInvoiceCreate"]

            if data.get("errors"):
                raise Exception(f"Blink error: {data['errors'][0]['message']}")

            inv = data["invoice"]
            exp_ms = int(time.time() * 1000) + 3_600_000
            macaroon = base64.b64encode(json.dumps({
                "hash": inv["paymentHash"],
                "exp": exp_ms,
            }).encode()).decode()

            return Invoice(
                payment_request=inv["paymentRequest"],
                payment_hash=inv["paymentHash"],
                macaroon=macaroon,
                amount_sats=amount_sats,
                expires_at=int(time.time()) + 3600,
            )

    async def check_payment(self, payment_hash: str) -> bool:
        query = """
        query CheckInvoice($paymentHash: PaymentHash!) {
          lnInvoice(paymentHash: $paymentHash) { status }
        }
        """
        async with httpx.AsyncClient() as client:
            res = await client.post(
                "https://api.blink.sv/graphql",
                headers={"X-API-KEY": self.api_key},
                json={"query": query, "variables": {"paymentHash": payment_hash}},
            )
            if not res.is_success:
                return False
            return res.json().get("data", {}).get("lnInvoice", {}).get("status") == "PAID"
